parley/src/tests/utils/env.rs

```diff
@@ -0,0 +1,221 @@
+        collection
+            .family_id(DEFAULT_FONT_NAME)
+            .unwrap_or_else(|| panic!("{} font not found", DEFAULT_FONT_NAME));
+        TestEnv {
+            test_name: test_name.to_string(),
+            check_counter: 0,
+            font_cx: FontContext {
+                collection,
+                source_cache: Default::default(),
+            },
+            tolerance: 0.0,
+            layout_cx: LayoutContext::new(),
+            foreground_color: Color::rgb8(0, 0, 0),
+            background_color: Color::rgb8(255, 255, 255),
+            errors: Vec::new(),
+        }
+    }
+
+    pub(crate) fn builder<'a>(&'a mut self, text: &'a str) -> RangedBuilder<'a, Color> {
+        let mut builder = self.layout_cx.ranged_builder(&mut self.font_cx, text, 1.0);
+        builder.push_default(StyleProperty::Brush(self.foreground_color));
+        builder.push_default(StyleProperty::FontStack(FontStack::Single(
+            FontFamily::Named(DEFAULT_FONT_NAME.into()),
+        )));
+        builder
+    }
+
+    fn image_name(&mut self) -> String {
+        let name = format!("{}-{}.png", self.test_name, self.check_counter);
+        self.check_counter += 1;
+        name
+    }
+
+    fn check_images(&self, current_img: &Pixmap, snapshot_path: &Path) -> Result<(), String> {
+        if !snapshot_path.is_file() {
+            return Err(format!("Cannot find snapshot {}", snapshot_path.display()));
+        }
+        let snapshot_img = Pixmap::load_png(snapshot_path)
+            .map_err(|_| format!("Loading snapshot {} failed", snapshot_path.display()))?;
+        if snapshot_img.width() != current_img.width()
+            || snapshot_img.height() != current_img.height()
+        {
+            return Err(format!(
+                "Snapshot has different size: snapshot {}x{}; generated image: {}x{}",
+                snapshot_img.width(),
+                snapshot_img.height(),
+                current_img.width(),
+                current_img.height()
+            ));
+        }
+
+        let mut n_different_pixels = 0;
+        let mut color_cumulative_difference = 0.0;
+        for (pixel1, pixel2) in snapshot_img.pixels().iter().zip(current_img.pixels()) {
+            if pixel1 != pixel2 {
+                n_different_pixels += 1;
+            }
+            let diff_r = (pixel1.red() as f32 - pixel2.red() as f32).abs();
+            let diff_g = (pixel1.green() as f32 - pixel2.green() as f32).abs();
+            let diff_b = (pixel1.blue() as f32 - pixel2.blue() as f32).abs();
+            color_cumulative_difference += diff_r.max(diff_g).max(diff_b);
+        }
+        if color_cumulative_difference > self.tolerance {
+            return Err(format!(
+                "Testing image differs in {n_different_pixels} pixels (color difference = {color_cumulative_difference})",
+            ));
+        }
+        Ok(())
+    }
+
+    pub(crate) fn check(&mut self, layout: &Layout<Color>) {
```

Comment:
I wonder if this could be renamed; check is a very passive name which doesn't suggest filesystem access.

Maybe just `screenshot` or `snapshot`?

Reply:
Renamed to `check_snapshot`.